File: codejam/cpp/2011/1A/1A_TheKillerWord.cpp
#include <bits/stdc++.h>
using namespace std;
typedef long long ll;
typedef vector<ll> vi;
typedef pair<ll,ll> pi;
#define FOR(i,a) for (ll i = 0; i < (a); i++)
#define len(x) (ll) x.size()
const ll INF = 1LL << 62;
const ll MOD = 1000000007;
//const ll MOD = 998244353;
const double PI = 4*atan(double(1.0));
vector<vector<ll>> twodi(ll N, ll M, ll v) { vector<vector<ll>> res(N); for (auto &vv : res) vv.resize(M,v); return res; }

int pat[10000][26];

int main (int argc, char **argv) {
    if (argc > 1) { freopen(argv[1],"r",stdin); }
    ios_base::sync_with_stdio(false);cin.tie(0);
    // PROGRAM STARTS HERE
    ll T; cin >> T;
    for (ll tt=1;tt<=T;tt++) {
        ll N,M; cin >> N >> M; vector<string> D(N); FOR(i,N) cin >> D[i]; vector<string> L(M); FOR(i,M) cin >> L[i];
        vector<vi> bylen(11);
        FOR(i,N) FOR(j,26) pat[i][j] = 0;
        FOR(i,N) {
            auto d = D[i];
            bylen[d.size()].push_back(i);
            FOR(j,d.size()) pat[i][d[j]-'a'] |= (1 << j);
        }
        vector<string> ansarr;
        FOR(lidx,M) {
            vi larr(26); FOR(i,26) larr[i] = L[lidx][i]-'a';
            ll bestidx(0), bestscore(-1);
            function<pair<ll,ll>(const vi&,ll)> dfs;
            dfs = [&](const vi& clist,ll lidx) -> pair<ll,ll> {
                ll l=larr[lidx];
                if (clist.size() == 1) return make_pair(0,clist[0]);
                map<ll,vi> lkup;
                for (auto &idx : clist) {
                    auto p = pat[idx][l];
                    if (lkup.count(p)==0) lkup[p] = vi();
                    lkup[p].push_back(idx); 
                }
                ll best(-1),bestidx(0);
                for (auto const& [key,val] : lkup) {
                    auto [lb,lbs] = dfs(val,lidx+1);
                    if (key == 0 && lkup.size() > 1) lb++;
                    if (lb > best || lb == best && lbs < bestidx) { best = lb; bestidx = lbs; }
                }
                return make_pair(best,bestidx);
            };
            for (ll s = 1; s <= 10; s++) {
                if (bylen[s].size() == 0) { continue; }
                auto [lb,lbs] = dfs(bylen[s],0);
                if (lb > bestscore || lb == bestscore && lbs < bestidx) { bestscore = lb; bestidx = lbs; }
            }
            ansarr.push_back(D[bestidx]);
        }
        cout << "Case #" << tt << ":";
        for (auto &a : ansarr) { cout << ' ' << a; }
        cout << '\n';
    }
}

File: codejam/python/2011/1A/1A_TheKillerWord.py
import sys
from collections import deque, namedtuple

## Input crap
infile = sys.stdin
intokens = deque()
def getTokens(): 
    while not intokens:
        for c in infile.readline().rstrip().split() :
            intokens.append(c)    
def gs(): getTokens(); return intokens.popleft()
def gi(): return int(gs())
def gf(): return float(gs())
def gbs(): return [c for c in gs()]
def gis(n): return [gi() for i in range(n)]
def ia(m): return [0] * m
def iai(m,v): return [v] * m
def twodi(n,m,v): return [iai(m,v) for i in range(n)]
def fill2(m) : r = gis(2*m); return r[0::2],r[1::2]
def fill3(m) : r = gis(3*m); return r[0::3],r[1::3],r[2::3]
def fill4(m) : r = gis(4*m); return r[0::4],r[1::4],r[2::4],r[3::4]
MOD = 998244353
##MOD = 1000000007

MyState = namedtuple("MyState","clist lidx score")
def main() :
    if len(sys.argv) > 1 : global infile; infile = open(sys.argv[1],'rt')
    ## PROGRAM STARTS HERE
    T = gi()
    for tt in range(1,T+1) :
        N,M = gi(),gi(); D = [gs() for _ in range(N) ]; L = [gs() for _ in range(M)]
        bylen = [[] for _ in range(11)]
        pat = [[0]*26 for _ in range(N)]
        for i,d in enumerate(D) :
            bylen[len(d)].append(i)
            for j,c in enumerate(d) :
                pat[i][ord(c)-ord('a')] |= 1 << j
        ansarr = []
        for lstr in L :
            larr = [ord(c)-ord('a') for c in lstr]
            bestidx,bestscore = 0,-1
            st = [MyState(a,0,0) for a in bylen if a]
            while st :
                lst = st.pop()
                if len(lst.clist) == 1 :
                    if lst.score > bestscore or lst.score == bestscore and bestidx > lst.clist[0] :
                        bestidx,bestscore = lst.clist[0],lst.score
                else :
                    d = {}; l = larr[lst.lidx]
                    for cc in lst.clist :
                        p = pat[cc][l]
                        if p in d : d[p].append(cc) 
                        else : d[p] = [cc]
                    for k,v in d.items() :
                        st.append(MyState(v,lst.lidx+1,lst.score + (1 if k==0 and len(d) > 1 else 0)))
            ansarr.append(D[bestidx])
        ans = " ".join(ansarr)
        print(f"Case #{tt}: {ans}")

if __name__ == "__main__" :
    main()

